File: src/switch_tracker/adapters/shopify.py
"""Shopify's public product feed.

Every Shopify storefront exposes its catalogue as JSON at /products.json with
no auth and no key.  Where collections are configured we page through those
instead -- faster, and politer than pulling a whole catalogue of unrelated
stock from a shop that also sells cameras.
"""

from __future__ import annotations

from typing import Any

from switch_tracker.adapters.base import ProgressSink
from switch_tracker.core.http import PoliteClient
from switch_tracker.core.models import (
    AdapterKind,
    Failed,
    FetchOutcome,
    Ok,
    Partial,
    Platform,
    ProductKind,
    RawListing,
    Region,
    StoreConfig,
)
from switch_tracker.core.parse import classify, infer_condition, infer_region, parse_price


class ShopifyAdapter:
    kind = AdapterKind.SHOPIFY

    PAGE_SIZE = 250
    # A backstop, NOT a target. The loop below already stops for real once a
    # page comes back short. This only bounds a genuinely broken feed that
    # never does, so it is generous rather than tuned to any one store's page
    # count -- a number like that goes stale the moment a catalogue grows.
    MAX_PAGES = 100

    def __init__(self, client: PoliteClient) -> None:
        self._client = client

    async def fetch(self, store: StoreConfig, sink: ProgressSink) -> FetchOutcome:
        paths = (
            [f"/collections/{c}/products.json" for c in store.collections]
            if store.collections
            else ["/products.json"]
        )

        listings: list[RawListing] = []
        problems: list[str] = []

        for path in paths:
            for page in range(1, self.MAX_PAGES + 1):
                url = f"{store.base_url}{path}?limit={self.PAGE_SIZE}&page={page}"
                data = await self._client.get_json(url)

                if not isinstance(data, dict):
                    problems.append(f"{path} page {page}: no parseable JSON")
                    break

                products = data.get("products") or []
                if not products:
                    break

                for raw in products:
                    listing = self._to_listing(store, raw)
                    if listing is not None:
                        listings.append(listing)

                sink.page(store.id, page, len(listings))

                # The real stopping signal: a short page means the last page.
                if len(products) < self.PAGE_SIZE:
                    break

        if not listings:
            return Failed("; ".join(problems) or "zero products returned")
        if problems:
            return Partial(tuple(listings), "; ".join(problems))
        return Ok(tuple(listings))

    def _to_listing(self, store: StoreConfig, product: dict[str, Any]) -> RawListing | None:
        variants = product.get("variants") or []
        if not variants:
            return None
        variant = variants[0]

        price = parse_price(variant.get("price"))
        if price is None:
            return None

        # Everything the store tells us, not just the title. A terse "Hogwarts
        # Legacy" with a product_type of "Nintendo Switch Games" is a Switch
        # game; passing the title alone is what made these stores look empty.
        context = " ".join(
            [
                str(product.get("title", "")),
                str(product.get("product_type") or ""),
                " ".join(str(t) for t in (product.get("tags") or [])),
            ]
        )
        result = classify(context, store.platform_hint)

        # Only cartridges belong in a cartridge price history.
        if result.kind is not ProductKind.GAME or result.platform is Platform.UNKNOWN:
            return None

        images = product.get("images") or []
        title = str(product.get("title", ""))
        sku = str(variant.get("sku") or "").strip() or str(product.get("id"))

        return RawListing(
            store_id=store.id,
            sku=sku,
            url=f"{store.base_url}/products/{product.get('handle')}",
            title=title,
            native_currency=store.currency,
            native_price=price,
            in_stock=variant.get("available") is not False,
            platform=result.platform,
            region=infer_region(title, Region.IN),
            condition=infer_condition(context),
            image_url=str(images[0].get("src")) if images else None,
        )

File: src/adapters/shopify.ts
import { getJson } from "../core/http.ts";
import { classify, inferRegion, parsePrice } from "../core/parse.ts";
import type { Adapter, FetchOutcome, RawListing, StoreConfig } from "../core/types.ts";

interface ShopifyVariant {
  id: number;
  title: string;
  price: string;
  available: boolean;
  sku?: string;
}

interface ShopifyProduct {
  id: number;
  title: string;
  handle: string;
  product_type?: string;
  tags?: string[];
  variants: ShopifyVariant[];
  images?: { src: string }[];
}

const PAGE_SIZE = 250;
// A backstop, not a target — the loop below already stops for real once a
// page comes back with fewer than PAGE_SIZE items. This just bounds a
// genuinely broken feed that never does, so it's generous rather than tuned
// to whatever any one store's real page count happens to be today.
const MAX_PAGES = 100;

/**
 * Shopify exposes the entire catalogue as JSON at /products.json with no auth
 * and no key. Where collections are configured we page through those instead —
 * faster, and politer than pulling a whole catalogue of unrelated stock.
 */
export const shopifyAdapter: Adapter = {
  kind: "SHOPIFY",

  async fetch(store: StoreConfig): Promise<FetchOutcome> {
    const paths = store.collections?.length
      ? store.collections.map((c) => `/collections/${c}/products.json`)
      : ["/products.json"];

    const listings: RawListing[] = [];
    const problems: string[] = [];

    for (const path of paths) {
      let page = 1;
      for (; page <= MAX_PAGES; page++) {
        const url = `${store.baseUrl}${path}?limit=${PAGE_SIZE}&page=${page}`;
        const data = await getJson<{ products?: ShopifyProduct[] }>(url);

        if (!data) {
          problems.push(`${path} page ${page}: no parseable JSON`);
          break;
        }
        const products = data.products ?? [];
        if (products.length === 0) break;

        for (const p of products) {
          const listing = toListing(store, p);
          if (listing) listings.push(listing);
        }
        if (products.length < PAGE_SIZE) break;
      }
    }

    if (listings.length === 0) {
      return { status: "failed", reason: problems.join("; ") || "zero products returned" };
    }
    if (problems.length > 0) {
      return { status: "partial", listings, reason: problems.join("; ") };
    }
    return { status: "ok", listings };
  },
};

function toListing(store: StoreConfig, p: ShopifyProduct): RawListing | undefined {
  const variant = p.variants?.[0];
  if (!variant) return undefined;

  const price = parsePrice(variant.price);
  if (price === undefined) return undefined;

  // Everything the store tells us, not just the title. A terse "Hogwarts
  // Legacy" in a product_type of "Nintendo Switch Games" is a Switch game;
  // passing the title alone is what made these stores look empty.
  const context = [p.title, p.product_type ?? "", (p.tags ?? []).join(" ")].join(" ");
  const { platform, kind } = classify(context, store.platformHint);

  // These shops also sell consoles, Joy-Cons, cases and eShop credit. Only
  // cartridges belong in a cartridge price history.
  if (kind !== "GAME" || platform === "UNKNOWN") return undefined;

  return {
    storeId: store.id,
    sku: variant.sku?.trim() || String(p.id),
    url: `${store.baseUrl}/products/${p.handle}`,
    title: p.title,
    nativeCurrency: store.currency,
    nativePrice: price,
    inStock: variant.available !== false,
    platform,
    region: inferRegion(p.title, "IN"),
    ...(p.images?.[0]?.src ? { imageUrl: p.images[0].src } : {}),
  };
}
